mcs/class/corlib/System.Reflection.Emit/ParameterBuilder.cs

```diff
@@ -96,12 +96,24 @@ public virtual void SetConstant (object defaultValue)
 		{
 			if (position > 0) {
 				Type t = methodb.GetParameterType (position - 1);
-				if (defaultValue != null && t != defaultValue.GetType ()) {
-					if(!t.IsEnum || t.UnderlyingSystemType != defaultValue.GetType ())
-						throw new ArgumentException ("Constant does not match the defined type.");
+				if (defaultValue == null) {
+					if (!(t.IsGenericType && t.GetGenericTypeDefinition () == typeof(Nullable<>)))
+						throw new ArgumentException("Null is not a valid constant value for this type");
```

Comment:
The Roslyn compilers can be made to encode a metadata constant of `null` for *any* parameter type, e.g. in this scenario:

```csharp
void Method<T>(T arg = default(T)) { }
```

Because the type of `T` is unknown at compile-time, the Roslyn compilers always encode `default(T)` as `null`.

It should be possible to do the same with Reflection.Emit. The above `if`s might collide with this particular scenario.

Reply:
Also, the Roslyn C# compiler generally encodes `default(SomeStruct)` as a `null` metadata constant:

```
struct SomeStruct { }

void Method(SomeStruct arg = default(SomeStruct)) { }
```

(`decimal` and `DateTime` are notable exceptions because they can be encoded in custom attributes, too, which the C# compiler does for `decimal` and the VB.NET compiler does for both types.)

Again, it should be possible to do the same with Reflection.Emit, and these `if`s might collide with that.

Basically, setting a default value of `null` should be allowed always. (I agree this is weird, but that's what the Roslyn compilers do today and what people might encounter via Reflection.)

Reply:
(Oh and if `t` is a reference type, `null` should definitely be an allowed default value. Looks like the above `if`s collide with that, too.)